x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -259,7 +265,16 @@ static Set<String> fieldNames(LogicalPlan parsed, Set<String> enrichPolicyMatchF
                 references.addAll(enrichRefs);
             } else {
                 references.addAll(p.references());
-                if (p instanceof Keep) {
+                // special handling for UnresolvedPattern (which is not an UnresolvedAttribute)
+                p.forEachExpression(UnresolvedNamePattern.class, up -> {
```

Comment:
Specifically for this change, please add some tests in `IndexResolverFieldNamesTests`.